karapace/schema_models.py

```diff
@@ -188,21 +229,41 @@ def parse(
     references: Sequence[Reference] | None = None,
     dependencies: Mapping[str, Dependency] | None = None,
     normalize: bool = False,
+    dependencies_compat: bool = False,
 ) -> ParsedTypedSchema:
     if schema_type not in [SchemaType.AVRO, SchemaType.JSONSCHEMA, SchemaType.PROTOBUF]:
         raise InvalidSchema(f"Unknown parser {schema_type} for {schema_str}")
-
+    parsed_schema_result: Draft7Validator | AvroSchema | ProtobufSchema
     parsed_schema: Draft7Validator | AvroSchema | ProtobufSchema
     if schema_type is SchemaType.AVRO:
         try:
+            if dependencies or dependencies_compat:
+                wrapped_schema_str = AvroMerge(schema_str, dependencies).wrap()
+            else:
+                wrapped_schema_str = schema_str
             parsed_schema = parse_avro_schema_definition(
-                schema_str,
+                wrapped_schema_str,
                 validate_enum_symbols=validate_avro_enum_symbols,
                 validate_names=validate_avro_names,
             )
+            if dependencies or dependencies_compat:
+                if isinstance(parsed_schema, avro.schema.UnionSchema):
+                    parsed_schema_result = parsed_schema.schemas[-1].fields[0].type.schemas[-1]
```

Comment:
Please what is going on here? Can you give some hint as to what you want to achieve, taking into account also the checks above?
```python
if dependencies or dependencies_compat:
    if isinstance(parsed_schema, avro.schema.UnionSchema):
```

Reply:
"To achieve schema merging with references (dependencies), we wrap the top-level schema and its references in a special way, then merge all schemas into a Union, where the top-level schema is the last, and all reference schemas are appended in a specific order with the rule that all referenced schemas must appear higher in the Union list than their referers (I mentioned this order in my answer about 'reverse' above). Then we parse this merged schema, resulting in an avro.schema.UnionSchema object, where the top-level schema is last in the union.

The AVRO object is constructed in such a way that dependent schemas are stored inside the AVRO object, and the top-level schema references them in a way that allows schema comparison without issues. However, this union also includes all dependency items at the same level as the top-level schema. So, we remove these items and extract only the top-level schema object from the union. All references remain present and work perfectly. As a result, we have no union, but we have a correct AVRO schema with dependencies.

However, there is a case where the top-level schema may have no references but is a Union itself. In this case, the code above checks if we have a UnionSchema as a result, and if this schema has references, we extract the top-level schema from it. If it is a Union without references, then we do nothing."